### ISSUE TITLE
How to cast nan to not-nan in an expression

### ISSUE THREAD

**Issue by zexuan-zhou** [sympy/sympy#26784]
Hi, I'm new to Sympy and I wonder how to cast null value to a default value in a Sympy expression
For example
```
import sympy as sp
x = sp.symbols("x")
# Return 0 if x is of nan value, and 1 otherwise
exp = sp.Piecewise((0.0, sp.Eq(x, sp.nan)), (1.0, True))
sp.lambdify((x), exp)(x=float('nan'))
Out[28]: 1.0
```
The above expression returns 1.0 instead of 0.0
I found that the problem comes from the expression
```
In [29]: sp.Eq(x, sp.nan)
Out[29]: False
```
It seems that `sp.Eq` does symbolic evaluation on variable `x` instead of waiting for the input from the actual calculation

Maybe I'm using sympy in a wrong way. But would appreciate someone to shed some light on how to do this

Thank you

**Comment by zexuan-zhou** [comment]
@oscarbenjamin If you could take a look. Thanks

**Comment by oscarbenjamin** [comment]
> In [29]: sp.Eq(x, sp.nan)
> Out[29]: False

I don't think this should give False but rather unknown or something.

**Comment by asmeurer** [comment]
Even if Eq were fixed to not evaluate, `lambdify` would still convert the code to `equal(x, nan)`, which is wrong. It should create `isnan(x)`.

**Comment by asmeurer** [comment]
The simplest workaround for you is to make your own `isnan` and use that instead of `Eq(x, nan)`. Currently this requires importing and using BooleanFunction, since Piecewise will fail if the expression is not a `Boolean`

```py
>>> from sympy.logic.boolalg import BooleanFunction
>>> class isnan(BooleanFunction):
...     pass
>>> isnan(x)
isnan(x)
>>> expr = Piecewise((0.0, isnan(x)), (1.0, True))
>>> lambdify((x), expr)(x=float('nan'))
array(0.)
```

This works because the function `isnan` uses the same name as the NumPy `isnan`, so when lambdified it just uses `np.isnan`. 

You could also add some logic to the custom `isnan` to actually evaluate to True on `sympy.nan` if you wanted.

**Comment by oscarbenjamin** [comment]
> `lambdify` would still convert the code to `equal(x, nan)`, which is wrong. It should create `isnan(x)`.

It is not clear to me that `Eq(x, nan)` should be evaluated as `isnan(x)`. The nan can come from other operations:
```
Eq(x, y/sin(y)).subs(y, 0)
```
Should substituting `x = nan` here return True?

**Comment by asmeurer** [comment]
The point is just that the semantics should remain the same after lambdifying. Sometimes SymPy and NumPy use different semantics and lambdify has to translate from one to the other. In this case, sympy.nan doesn't use the same equality logic as numpy.nan, so lambdify should translate Eq to the equivalent isnan when one of the operands is nan. 

sympy.nan is already the codegen version of numpy.nan, even if they don't necessarily represent the same things. 

Now lambdify could in principle always translate `Eq(x, y)` to `np.where(isnan(x) | isnan(y), False, x == y)` but I don't think we should do that. Trying to handle nan in every possible function would just result in making everything slower, but it should at least handle it when an explicit `sympy.nan` is present.

**Comment by oscarbenjamin** [comment]
> lambdify should translate Eq to the equivalent isnan when one of the operands is nan

Sympy's Eq does not behave like `isnan` so there is no equivalence.

**Comment by asmeurer** [comment]
The equivalent to `Piecewise((0.0, sp.Eq(x, sp.nan)), (1.0, True))` uses `isnan`. In SymPy, replacing `x` with `nan` produces `0.0`, so in NumPy it should do the same. Like I said, `sympy.nan` is the codegen version of `np.nan`, so if you want to add nan handling in your expression so that it would get translated to lambdify, that's how you would do it. 

I guess we could also add `isnan` to sympy.codegen, but I in general if some SymPy function has a certain behavior that they want to translate to NumPy we want people to just that.

**Comment by oscarbenjamin** [comment]
> The equivalent to `Piecewise((0.0, sp.Eq(x, sp.nan)), (1.0, True))` uses `isnan`

Can you demonstrate the equivalence here? We have:
```python
In [12]: Eq(nan, nan)
Out[12]: False

In [13]: np.isnan(float('nan'))
Out[13]: True
```
Do you propose that `Eq(nan, nan)` should evaluate to True?

**Comment by oscarbenjamin** [comment]
> The equivalent to `Piecewise((0.0, sp.Eq(x, sp.nan)), (1.0, True))` uses `isnan`. In SymPy, replacing `x` with `nan` produces `0.0`, so in NumPy it should do the same.

This isn't what happens:
```python
In [15]: Piecewise((0.0, sp.Eq(x, sp.nan)), (1.0, True))
Out[15]: 1.00000000000000

In [16]: Piecewise((0.0, sp.Eq(sp.nan, sp.nan)), (1.0, True))
Out[16]: 1.00000000000000

In [17]: Piecewise((0.0, sp.Eq(x, sp.nan)), (1.0, True)).subs(x, sp.nan)
Out[17]: 1.00000000000000
```
No cases give 0.0.

**Comment by zexuan-zhou** [comment]
Thanks guys. I looked at BooleanFunction and it seems not very trivial for me to implement a customized version of `isnan`
Also a FYI, my use case is with tensorflow as the backend compute module when using `lambdify` so having something that only works in numpy is not sufficient

**Comment by bjodah** [comment]
There is a `isnan` function in `sympy.codegen.cfunctions`. But it looks like it should be subclassing booleanfunction becuase Piecewise currently does not seem to work with it:
<details>
<summary><pre>>>> f = lambdify(x, Piecewise((0.0, isnan(x)), (1.0, True)))</pre></summary>

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 f = lambdify(x, Piecewise((0.0, isnan(x)), (1.0, True)))

File ~/vc/sympy/sympy/functions/elementary/piecewise.py:137, in Piecewise.__new__(cls, *args, **options)
    134 newargs = []
    135 for ec in args:
    136     # ec could be a ExprCondPair or a tuple
--> 137     pair = ExprCondPair(*getattr(ec, 'args', ec))
    138     cond = pair.cond
    139     if cond is false:

File ~/vc/sympy/sympy/functions/elementary/piecewise.py:33, in ExprCondPair.__new__(cls, expr, cond)
     30         cond = cond.rewrite(ITE)
     32 if not isinstance(cond, Boolean):
---> 33     raise TypeError(filldedent('''
     34         Second argument must be a Boolean,
     35         not `%s`''' % func_name(cond)))
     36 return Tuple.__new__(cls, expr, cond)
```

</details>

```
TypeError:
Second argument must be a Boolean, not `isnan`
```

**Comment by asmeurer** [comment]
> Do you propose that Eq(nan, nan) should evaluate to True?

Oh I didn't realize Eq was implementing that behavior. That's why Eq(x, nan) evaluates to False apparently, which is what you suggested should be changed above. 

Personally I would be OK with SymPy's nan not carrying over the weird IEEE nan equality stuff. Really SymPy's nan was a poor decision and it should have just been called "undefined" or something.

**Comment by zexuan-zhou** [comment]
Would it be possible to fix this soon? Thanks

**Comment by oscarbenjamin** [comment]
It is not completely clear to me what the fix should be. I think probably the code in the OP should not be expected to work and should be written differently.

**Comment by bjodah** [comment]
@zexuan-zhou I wonder if you can try gh-26794 and see if that works for you.

**Comment by zexuan-zhou** [comment]
@bjodah Thanks. I tested it and it seems work now in regular mode but not in tensorflow

```
In [4]: !git branch
* add-isinf
  master

In [5]: from sympy.codegen.cfunctions import isnan

In [6]: import sympy as sp
   ...: x = sp.symbols("x")
   ...: # Return 0 if x is of nan value, and 1 otherwise
   ...: exp = sp.Piecewise((0.0, isnan(x)), (1.0, True))
   ...: sp.lambdify((x), exp)(x=float('nan'))
Out[6]: array(0.)

In [7]: sp.lambdify((x), exp)(x=1)
Out[7]: array(1.)

In [13]: sp.lambdify((x), exp, modules="tensorflow")(x=tf.constant([np.nan]))
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[13], line 1
----> 1 sp.lambdify((x), exp, modules="tensorflow")(x=tf.constant([np.nan]))
File <lambdifygenerated-6>:2, in _lambdifygenerated(x)
      1 def _lambdifygenerated(x):
----> 2     return where(isnan(x), 0.0, 1.0)

NameError: name 'isnan' is not defined
```

**Comment by bjodah** [comment]
Yes, that's expected. Feel free to add a print method to the tensorflow printer (I'm not familiar with tensorflow). I can pull in a commit into this branch if you would like.

**Comment by zexuan-zhou** [comment]
@bjodah Would appreciate it if you could help. I'm not familiar with how to modify the printer but I can give it a try if there is a guide

**Comment by bjodah** [comment]
Sure, this functions looks like it does what we want:
https://www.tensorflow.org/api_docs/python/tf/math/is_nan

Then if you search for "def _print_isnan" in the diff view of my PR: https://github.com/sympy/sympy/pull/26794/files

You'll see how to implement the print method. You will need to do this here:
https://github.com/sympy/sympy/blob/master/sympy/printing/tensorflow.py

**Comment by zexuan-zhou** [comment]
@bjodah Thanks!
Are you going to merge your PR? I can rebase once you merge

**Comment by bjodah** [comment]
Typically we do not self-merge PRs in the sympy project. So we'll have to wait until someone finds the time to review the PR. You can open a new PR once mine is merged, or you can even open a PR against my branch over at github.com/bjodah/sympy if you prefer. It's up to you. :)
